tutorials/math/combinatorics.md

```diff
@@ -1,10 +1,215 @@
+
+## Finding $n \choose r$
+
+Pronounced "n choose r", this is the mathematical notation to represent number of ways to choose r objects out of a collection of n objects. The analytical formula can be written as $n \choose r$ $=$ $\frac{n!}{r!\,(n - r)!}$.
+
+This leads to a neat recurrence relation :-  $n \choose r$ $=$ $n - 1 \choose r$ + $n - 1 \choose r - 1$ $\\$
+We can precompute all the required values using the above formula in $O(n^2)$ and then perform lookup in $O(1)$ time. The resulting values also form [Pascal's Triangle](https://en.wikipedia.org/wiki/Pascal%27s_triangle).
+
+### Example #1 [2221 - Find Triangular Sum of an Array](https://leetcode.com/problems/find-triangular-sum-of-an-array/)
+
+The important insight here is that the figure provided is nothing but an inverted Pascal's Triangle and contribution of each cell in the final sum is $value\,of\,cell$ $*$ $binomial\,coefficient\,at\,the\,particular\,position\,in\,Pascal's\,Triangle$
+
+Thus for the cell at $i^{th}$ index in the topmost row, it's value is multiplied by $n - 1 \choose i$ and added to the final sum $modulo\,10$.
+Time Complexity of the program is $O(n^2)$ for computing the binomial coefficient and $O(n)$ Space complexity.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    int triangularSum(vector<int>& nums) {
+        int n = nums.size();
+        vector<int> pascalTriangleRow = {1};
+        // Calculate the ith row using (i - 1)th row
+        for(int i = 0; i < n; i++){
+            vector<int> nextRow = {1};
+            for(int j = 1; j < i; j++){
+                nextRow.push_back((pascalTriangleRow[j] + pascalTriangleRow[j - 1])%10);
+            }
+            nextRow.push_back(1);
+            pascalTriangleRow = nextRow;
+        }
+        // Calculate the final answer as discussed above
+        int ans = 0;
+        for(int i = 0; i < n; i++){
+            ans += (nums[i]*pascalTriangleRow[i])%10;
+        }
+        return ans%10;
+    }
+};
+```
+
+</TabItem>
+</Tabs>
+
+Sometimes it is not possible to calculate the entirety of Pascal's Triangle due to larger values of $n$. In this case, we begin by precomputing $x!$   $\forall$ $x \in [{0, n}]$. Similarly, we will also [precompute](https://cp-algorithms.com/algebra/module-inverse.html#mod-inv-all-num) the modular inverses. This can be achieved in $O(n)$ time. Thus we can now compute $n \choose r$ using the analytical equation presented earlier. You can read about modular inverses [here](../../tutorials/basic-topics/mod.md)
+
+The implementation of above can be as follows:-
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
```

Comment:
1. add space. see above.
2. Using this template with your example below in the problem 62 - Unique Paths produces wrong answer.